gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationFragment.java

```diff
@@ -119,116 +147,107 @@ private void handleSaveResult(EditObservationViewModel.SaveResult saveResult) {
     }
   }
 
-  private void rebuildForm(Form form) {
-    formLayout.removeAllViews();
-    for (Element element : form.getElements()) {
-      switch (element.getType()) {
-        case FIELD:
-          addField(element.getField());
-          break;
-        default:
-          Timber.d("%s elements not yet supported", element.getType());
-          break;
-      }
-    }
-  }
+  private <V extends AbstractFieldViewModel> V addFieldViewModel(Field field) {
+    V fieldViewModel = (V) viewModelFactory.create(getViewModelClass(field.getType()));
 
-  private void addField(Field field) {
-    switch (field.getType()) {
-      case TEXT:
-        addTextField(field);
-        break;
-      case MULTIPLE_CHOICE:
-        addMultipleChoiceField(field);
-        break;
-      case PHOTO:
-        addPhotoField(field);
-        break;
-      default:
-        Timber.w("Unimplemented field type: %s", field.getType());
-        break;
+    // load field and current response
+    fieldViewModel.init(field, viewModel.getResponse(field.getId()));
+
+    // handle UI interactions
+    if (fieldViewModel instanceof PhotoFieldViewModel) {
+      observeSelectPhotoClicks((PhotoFieldViewModel) fieldViewModel);
+    } else if (fieldViewModel instanceof MultipleChoiceFieldViewModel) {
+      observeMultipleChoiceClicks((MultipleChoiceFieldViewModel) fieldViewModel);
     }
-  }
 
-  private void addTextField(Field field) {
-    TextInputFieldBinding binding =
-        TextInputFieldBinding.inflate(getLayoutInflater(), formLayout, false);
-    binding.setViewModel(viewModel);
-    binding.setLifecycleOwner(this);
-    binding.setField(field);
-    formLayout.addView(binding.getRoot());
-    assignGeneratedId(binding.getRoot().findViewById(R.id.text_input_edit_text));
-  }
+    // subscribe to response updates
```

Comment:
Style nit: Please use capital case with final period in comments. Example: "Please use capital case with final period in comments." :)

Reply:
Also please make sure the comment describes something more than what's clear from the code itself. In this case, for example, a comment isn't needed.